### PR TITLE
Fix compilation for multi-configuration generators and debug builds

### PR DESCRIPTION
**Pull request by jamiesnape**
TLDR: Previous approach assumed PODs.

The previous approach was assuming a single configuration build using Makefiles per the PODs conventions. This caused issues with multi-configuration builds in particular, but also debug builds, mostly because it did not take into account that CMake would add its own flags to distinguish between Debug, Release, and other build configurations. Those flags would partially conflict with those added using the introspection of the MEX compilation.

Fortunately, using the optimization flags provided by CMake works well with the MEX compiler (see also here and CMake upstream for the function [matlab_add_mex](https://github.com/Kitware/CMake/blob/master/Modules/FindMatlab.cmake#L790-L1044) in [FindMatlab.cmake](https://github.com/Kitware/CMake/blob/master/Modules/FindMatlab.cmake)). The best approach would be to use that function, which also makes `mex_setup` redundant, but, for now, this fixes compilation.

Also Drake overwrote a flags in several places and there were some missing POD installs, so it is surprising that any configuration worked.

Not upstreaming (yet) because, as far as I can tell, any other consumers are using PODs.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/3154)

### CODE REVIEW

File: cmake/mex.cmake [hunk 1]
```diff
@@ -225,21 +225,15 @@ function(mex_setup)
 
 
   # todo: handle C separately from CXX?
-  set (MEX_COMPILE_FLAGS "${MEX_INCLUDE} ${MEX_CXXFLAGS} ${MEX_DEFINES} ${MEX_MATLABMEX} ${MEX_CXX_ARGUMENTS}")
```
Comment (jamiesnape) [inline cmake/mex.cmake]: MEX_CXXFLAGS includes MEX_DEFINES on all platforms, which includes MEX_MATLABMEX. Another was added below, so at the very least there were four `-DMATLAB_MEX_FILE` flags on each target...

File: cmake/mex.cmake [hunk 2]
```diff
@@ -225,21 +225,15 @@ function(mex_setup)
...
     endif()
-  elseif (_build_type MATCHES RELEASE)
-    set(MEX_COMPILE_FLAGS "${MEX_CXXOPTIMFLAGS} ${MEX_COMPILE_FLAGS}")
   endif()
 
-  if (${MEX_COMPILE_FLAGS} MATCHES "-ansi")
-    string(REPLACE "-ansi" "" MEX_COMPILE_FLAGS "${MEX_COMPILE_FLAGS}")
-    message(WARNING "Your MEX compiler flags contained '-ansi', but we've removed that flag for compatibility with C++11")
-  endif()
+  string(REPLACE "-ansi" "" MEX_COMPILE_FLAGS "${MEX_COMPILE_FLAGS}")
```
Comment (jamiesnape) [inline cmake/mex.cmake]: Technically this does not affect compatibility with C++11 on supported platforms, though it would affect C11 compatibility.

**Comment by jamiesnape** [comment]
@drake-jenkins-bot linux-gcc-ninja-experimental-matlab please
@drake-jenkins-bot linux-gcc-experimental-matlab-open-source please
@drake-jenkins-bot linux-gcc-ninja-experimental-matlab-open-source please
@drake-jenkins-bot linux-gcc-experimental-coverage-matlab please
@drake-jenkins-bot windows-msvc-32-experimental-matlab please
@drake-jenkins-bot windows-msvc-ninja-32-experimental-matlab please
@drake-jenkins-bot windows-msvc-64-experimental-matlab please
@drake-jenkins-bot windows-msvc-ninja-64-experimental-matlab please

**Comment by jamiesnape** [comment]
https://drake-jenkins.csail.mit.edu/job/windows-msvc-32-experimental-matlab/22/
https://drake-jenkins.csail.mit.edu/job/windows-msvc-64-experimental-matlab/35/

**Comment by mwoehlke-kitware** [comment]
> Not upstreaming (yet) because, as far as I can tell, any other consumers are using PODs.

As discussed in the meeting, I'd rather get us eventually ported to `FindMatlab.cmake` anyway, and in the short term I want to get `mex.cmake` pulled *out* of the RobotLocomotion/cmake repo. That said, I'm not sure there *are* other consumers of `mex.cmake`... (IOW, I agree with the above but would leave out the "yet" :smile:.)

> Also importing `FindMatlab.cmake` [...]

If you're going to do this, it belongs in `cmake/modules/3.7` (and preferably we should get changes into upstream *first*, before incorporating them locally). Is this PR making use of it in any way yet, though? If not (and offhand it looks to me like it isn't), I'd prefer to just leave it out for now.

**Comment by jamiesnape** [comment]
> If you're going to do this, it belongs in cmake/modules/3.7 (and preferably we should get changes into upstream first, before incorporating them locally). Is this PR making use of it in any way yet, though? If not (and offhand it looks to me like it isn't), I'd prefer to just leave it out for now.

I guess I misheard you in the meeting. Gone.

**Comment by mwoehlke-kitware** [comment]
> I guess I misheard you in the meeting.

I don't recall for certain what I said. Definitely having `mex.cmake` local will help with #3065. I know we talked about upstreaming changes to `FindMatlab.cmake`, and I thought I also mentioned "how to include from CMake future the right way", but "the right way" isn't merged or even PR'd yet (and I'm resisting the urge to point to a commit in my fork because the commit I have right now is temporary and is going to become a dead link as soon as [cmake!63](https://gitlab.kitware.com/cmake/cmake/merge_requests/63) lands and I update the branch).

**Comment by jamiesnape** [comment]
@drake-jenkins-bot retest this please

**Comment by jamiesnape** [comment]
@drake-jenkins-bot linux-gcc-experimental-matlab

**Comment by jamiesnape** [comment]
@drake-jenkins-bot windows-msvc-32-experimental-matlab

**Comment by jamiesnape** [comment]
@drake-jenkins-bot windows-msvc-32-experimental-matlab please

**Comment by jamiesnape** [comment]
@drake-jenkins-bot windows-msvc-ninja-32-experimental-matlab please

**Comment by mwoehlke-kitware** [comment]
>  I thought I also mentioned "how to include from CMake future the right way", but "the right way" isn't merged or even PR'd yet

Now it is: see #3174. Anything in CMake upstream (e.g. `FindMatlab.cmake`) should be included following that pattern. (And ideally after that PR lands, since it has the logic to set up `CMAKE_MODULE_PATH`.)

Note also that "following that pattern" implies that *upstream has accepted the changes*, though I suppose we could drop stuff in e.g. `modules/99999` if we can't wait for that (and move it to the right spot later).

**Comment by jamiesnape** [comment]
@mwoehlke-kitware I am having issues with compiler flags disappearing on this PR. Should these commands have an `APPEND` appended?

https://github.com/RobotLocomotion/cmake/blob/master/pods.cmake#L516-L517

**Comment by mwoehlke-kitware** [comment]
> Should [these commands](https://github.com/RobotLocomotion/cmake/blob/master/pods.cmake#L516-L517) have an `APPEND` appended?

...probably? (Or `APPEND_STRING`, not sure which is appropriate with those properties.)

However, I don't think I've ever needed those; in other projects I've only ever just linked to `${CMAKE_THREAD_LIBS_INIT}`... maybe they can just go away entirely? (Even if not, we could — and probably should — just fix `FindThreads.cmake` to produce an `IMPORTED` or `INTERFACE` target with appropriate usage requirements, which will also make that logic unnecessary.)

**Comment by billhoffman** [comment]
It was always my plan to upstream the changes in FindMatlab.cmake that are in drake.   That should be done sooner than later.

**Comment by jamiesnape** [comment]
@drake-jenkins-bot linux-gcc-experimental-matlab-open-source please
@drake-jenkins-bot linux-gcc-ninja-experimental-matlab please
@drake-jenkins-bot linux-gcc-ninja-experimental-matlab-open-source please
@drake-jenkins-bot mac-clang-experimental-matlab-please
@drake-jenkins-bot windows-msvc-32-experimental-matlab please
@drake-jenkins-bot windows-msvc-64-experimental-matlab please
@drake-jenkins-bot windows-msvc-ninja-32-experimental-matlab please
@drake-jenkins-bot windows-msvc-ninja-64-experimental-matlab please

**Comment by jamiesnape** [comment]
Debug builds:
* https://drake-jenkins.csail.mit.edu/job/windows-msvc-32-experimental-matlab/44/
* https://drake-jenkins.csail.mit.edu/job/windows-msvc-64-experimental-matlab/51/
* https://drake-jenkins.csail.mit.edu/job/windows-msvc-ninja-32-experimental-matlab/34/
* https://drake-jenkins.csail.mit.edu/job/windows-msvc-ninja-64-experimental-matlab/48/

**Comment by jamiesnape** [comment]
@drake-jenkins-bot linux-gcc-experimental-matlab-open-source please
@drake-jenkins-bot linux-gcc-ninja-experimental-matlab please
@drake-jenkins-bot linux-gcc-ninja-experimental-matlab-open-source please
@drake-jenkins-bot mac-clang-experimental-matlab-please
@drake-jenkins-bot windows-msvc-32-experimental-matlab please
@drake-jenkins-bot windows-msvc-64-experimental-matlab please
@drake-jenkins-bot windows-msvc-ninja-32-experimental-matlab please
@drake-jenkins-bot windows-msvc-ninja-64-experimental-matlab please

**Comment by jamiesnape** [comment]
@drake-jenkins-bot linux-gcc-experimental-matlab-open-source please
@drake-jenkins-bot linux-gcc-ninja-experimental-matlab please
@drake-jenkins-bot linux-gcc-ninja-experimental-matlab-open-source please
@drake-jenkins-bot mac-clang-experimental-matlab-please
@drake-jenkins-bot windows-msvc-32-experimental-matlab please
@drake-jenkins-bot windows-msvc-64-experimental-matlab please
@drake-jenkins-bot windows-msvc-ninja-32-experimental-matlab please
@drake-jenkins-bot windows-msvc-ninja-64-experimental-matlab please

**Comment by jamiesnape** [comment]
@drake-jenkins-bot windows-msvc-ninja-64-experimental-matlab please
@drake-jenkins-bot mac-clang-experimental-matlab-please

**Comment by jamiesnape** [comment]
@drake-jenkins-bot windows-msvc-64-experimental-matlab please

**Comment by jamiesnape** [comment]
@drake-jenkins-bot linux-gcc-experimental-matlab-open-source please
@drake-jenkins-bot linux-gcc-ninja-experimental-matlab please
@drake-jenkins-bot linux-gcc-ninja-experimental-matlab-open-source please
@drake-jenkins-bot windows-msvc-ninja-32-experimental-matlab please
@drake-jenkins-bot windows-msvc-ninja-64-experimental-matlab please
@drake-jenkins-bot windows-msvc-32-experimental-matlab please
@drake-jenkins-bot windows-msvc-64-experimental-matlab please

**Comment by jamiesnape** [comment]
@drake-jenkins-bot windows-msvc-ninja-32-experimental-matlab please

**Comment by jamiesnape** [comment]
@mwoehlke-kitware for feature review, @david-german-tri for platform review.

**Comment by mwoehlke-kitware** [comment]
Review status: 0 of 9 files reviewed at latest revision, 3 unresolved discussions.

---

*[cmake/mex.cmake, line 242 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3154#-KPrfcn8a7J7NhZYfXw4:-KPrfcn9hAF2XrbKL_G0:b-q98b3f) ([raw file](https://github.com/robotlocomotion/drake/blob/0ee9497240193df2160dc86dd1f8fa4527a9aaf0/cmake/mex.cmake#L242)):*
> ```CMake
>     # Controlled by CMAKE_C_STANDARD and CMAKE_CXX_STANDARD
>     string(REPLACE "-ansi " "" MEX_COMPILE_FLAGS "${MEX_COMPILE_FLAGS}")
>     string(REPLACE "-std=c++11 " "" MEX_COMPILE_FLAGS "${MEX_COMPILE_FLAGS}")
> ```

For future-proofing (or for that matter, in case it should happen to use 0x instead of 11), please make this a regex to replace any `-std=`. (Or mention here, and possible in the file in a comment, why this should not be done.)

---

*[drake/systems/plants/CMakeLists.txt, line 59 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3154#-KPridEnbisXCEMeMkbe:-KPridEo2BEOSuu2t66h:bz8sypu) ([raw file](https://github.com/robotlocomotion/drake/blob/0ee9497240193df2160dc86dd1f8fa4527a9aaf0/drake/systems/plants/CMakeLists.txt#L59)):*
> ```CMake
> if(MATLAB_FOUND)
>   add_mex(rbmMexFunctions SHARED rigidBodyTreeMexFunctions.cpp)
>   pods_install_libraries(rbmMexFunctions)
> ```

Should this be in a separate commit?

---

*[drake/systems/plants/constraint/CMakeLists.txt, line 20 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3154#-KPrimnMPO98UqLoEh-L:-KPrimnNBJ-FRZEnuvGC:bvocuaf) ([raw file](https://github.com/robotlocomotion/drake/blob/0ee9497240193df2160dc86dd1f8fa4527a9aaf0/drake/systems/plants/constraint/CMakeLists.txt#L20)):*
> ```CMake
>   add_mex(drakeConstructRigidBodyConstraint SHARED constructPtrRigidBodyConstraint.cpp)
>   target_link_libraries(drakeConstructRigidBodyConstraint drakeRigidBodyConstraint drakeRBM drakeMexUtil drakeGeometryUtil)
>   pods_install_libraries(drakeConstructRigidBodyConstraint)
> ```

likewise (possibly the same commit as the previous instance)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3154)*

**Comment by jamiesnape** [comment]
Review status: 0 of 9 files reviewed at latest revision, 3 unresolved discussions.

---

*[cmake/mex.cmake, line 242 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3154#-KPrfcn8a7J7NhZYfXw4:-KPrjUvY-kaTZPZNZPjx:bz24s3c) ([raw file](https://github.com/robotlocomotion/drake/blob/0ee9497240193df2160dc86dd1f8fa4527a9aaf0/cmake/mex.cmake#L242)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary>
> For future-proofing (or for that matter, in case it should happen to use 0x instead of 11), please make this a regex to replace any `-std=`. (Or mention here, and possible in the file in a comment, why this should not be done.)
</details>
These are not arbitrary flags, we basically know what they are now and what they have been in the past, so it will never be 0x.

---

*[drake/systems/plants/CMakeLists.txt, line 59 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3154#-KPridEnbisXCEMeMkbe:-KPrjAvwqU9GUHHZKLY4:b2nhky0) ([raw file](https://github.com/robotlocomotion/drake/blob/0ee9497240193df2160dc86dd1f8fa4527a9aaf0/drake/systems/plants/CMakeLists.txt#L59)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary>
> Should this be in a separate commit?
</details>
No.

---

*[drake/systems/plants/constraint/CMakeLists.txt, line 20 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3154#-KPrimnMPO98UqLoEh-L:-KPrjFLh-z0n73IYK8iL:b-fp9u8j) ([raw file](https://github.com/robotlocomotion/drake/blob/0ee9497240193df2160dc86dd1f8fa4527a9aaf0/drake/systems/plants/constraint/CMakeLists.txt#L20)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary>
> likewise (possibly the same commit as the previous instance)
</details>
Also, no.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3154)*

**Comment by jamiesnape** [comment]
I am not going to make any changes to comments until @david-german-tri has reviewed since I would rather not have to run 10+ builds again.

**Comment by mwoehlke-kitware** [comment]
Review status: 0 of 9 files reviewed at latest revision, 3 unresolved discussions.

---

*[cmake/mex.cmake, line 242 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3154#-KPrfcn8a7J7NhZYfXw4:-KPrnO16GPjixqHtUgWE:b-lagab5) ([raw file](https://github.com/robotlocomotion/drake/blob/0ee9497240193df2160dc86dd1f8fa4527a9aaf0/cmake/mex.cmake#L242)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary>
> These are not arbitrary flags, we basically know what they are now and what they have been in the past, so it will never be 0x.
</details>
...but we don't know what they'll be in the future. If there is a reason we want that `MEX_COMPILE_FLAGS` should possibly "suddenly" contain e.g. `-std=c++14`, I would feel better if there was a comment to that effect.

---

*[drake/systems/plants/CMakeLists.txt, line 59 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3154#-KPridEnbisXCEMeMkbe:-KPrn1sRqVZoJfnoc0hF:b-fk5y1z) ([raw file](https://github.com/robotlocomotion/drake/blob/0ee9497240193df2160dc86dd1f8fa4527a9aaf0/drake/systems/plants/CMakeLists.txt#L59)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary>
> No.
</details>
Why not? What does installing a library (which presumably was by accident not being installed before) have to do with fixing compilation issues?

Maybe it's fine and really does belong, but if so, I don't understand how/why.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3154)*

**Comment by jamiesnape** [comment]
Review status: 0 of 9 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/plants/CMakeLists.txt, line 59 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3154#-KPridEnbisXCEMeMkbe:-KPrpG3I-SfAjWE4Hx2n:b-7p9s6h) ([raw file](https://github.com/robotlocomotion/drake/blob/0ee9497240193df2160dc86dd1f8fa4527a9aaf0/drake/systems/plants/CMakeLists.txt#L59)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary>
> Why not? What does installing a library (which presumably was by accident not being installed before) have to do with fixing compilation issues?
> 
> Maybe it's fine and really does belong, but if so, I don't understand how/why.
</details>
I guess I am semi-abusing the word compilation since MATLAB is involved. The MEX needs to be in a certain place otherwise the MATLAB code will at runtime or JIT compilation, however MATLAB works.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3154)*

**Comment by jamiesnape** [comment]
Review status: 0 of 9 files reviewed at latest revision, 3 unresolved discussions.

---

*[cmake/mex.cmake, line 242 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3154#-KPrfcn8a7J7NhZYfXw4:-KPrpv7SqucscX9aSHhf:b19hsr) ([raw file](https://github.com/robotlocomotion/drake/blob/0ee9497240193df2160dc86dd1f8fa4527a9aaf0/cmake/mex.cmake#L242)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary>
> ...but we don't know what they'll be in the future. If there is a reason we want that `MEX_COMPILE_FLAGS` should possibly "suddenly" contain e.g. `-std=c++14`, I would feel better if there was a comment to that effect.
</details>
I will fix it later. All the links to the various builds will disappear the moment I push another commit, and in any case we only support one version of MATLAB.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3154)*

**Comment by mwoehlke-kitware** [comment]
Looks okay to me aside from the minor nits.

FWIW, IMHO it is not necessary to re-run CI builds if the PR content has not changed. However, I take your point about the links disappearing.

---

Review status: 0 of 9 files reviewed at latest revision, 3 unresolved discussions.

---

*[cmake/mex.cmake, line 242 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3154#-KPrfcn8a7J7NhZYfXw4:-KPrqt3ufu-Xz6xal4fg:b-nmz1t6) ([raw file](https://github.com/robotlocomotion/drake/blob/0ee9497240193df2160dc86dd1f8fa4527a9aaf0/cmake/mex.cmake#L242)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary>
> I will fix it later. All the links to the various builds will disappear the moment I push another commit, and in any case we only support one version of MATLAB.
</details>
Okay, thanks (and I agree with the reason for waiting).

---

*[drake/systems/plants/CMakeLists.txt, line 59 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3154#-KPridEnbisXCEMeMkbe:-KPrr-UFEh9Zavkcq3mk:bjpb7z1) ([raw file](https://github.com/robotlocomotion/drake/blob/0ee9497240193df2160dc86dd1f8fa4527a9aaf0/drake/systems/plants/CMakeLists.txt#L59)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary>
> I guess I am semi-abusing the word compilation since MATLAB is involved. The MEX needs to be in a certain place otherwise the MATLAB code will at runtime or JIT compilation, however MATLAB works.
</details>
I see. I'd still prefer it in a separate commit (same PR is fine though!), or at least mention it in the commit message. (For that matter, I wouldn't mind a more detailed commit message in general...)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3154#-:-KPrrEOqqL6OuNprXvCh:b4fy0kn)*

**Comment by david-german-tri** [comment]
Awesome!  Clearly this fixes the build, so :lgtm: in the hopes of getting it in today.  However, I do not understand how it works, at all.  Since you've taken the time to understand mex.cmake, could you add some comments?  See specific requests below.

---

Review status: 0 of 9 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[cmake/mex.cmake, line 132 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3154#-KPtC1-BS5skBITs3gBV:-KPtC1-CS3x4pMN9BrIv:b-ghu20m) ([raw file](https://github.com/robotlocomotion/drake/blob/0ee9497240193df2160dc86dd1f8fa4527a9aaf0/cmake/mex.cmake#L132)):*
> ```CMake
>     get_mex_option(CFLAGS NAMES COMPFLAGS)
>     get_mex_option(CXXFLAGS NAMES COMPFLAGS)
>     get_mex_option(DEFINES NAMES COMPDEFINES)
> ```

Is `get_mex_option` the "introspection" you've previously mentioned?  Could you add a comment to it explaining what it actually does?  (It looks like it assumes `mexv_output` is already in scope and parses it for these particular strings, or something like that?)

---

*[cmake/mex.cmake, line 231 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3154#-KPtCSr1oPtWeB4ZCINU:-KPtCSr1oPtWeB4ZCINV:b-ioqed8) ([raw file](https://github.com/robotlocomotion/drake/blob/0ee9497240193df2160dc86dd1f8fa4527a9aaf0/cmake/mex.cmake#L231)):*
> ```CMake
>   set (MEX_COMPILE_FLAGS "${MEX_CXXFLAGS} ${MEX_DEFINES} ${MEX_CXX_ARGUMENTS}")
> 
>   # CMake will add the appropriate flags for the chosen build configuration
> ```

Meaning, the mex compiler suggested some flags, and we are editing them out so that CMake can set them instead?  Could you elaborate this comment a little?

---

*[cmake/mex.cmake, line 317 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3154#-KPtCadmmq1pu_lOGMiU:-KPtCadmmq1pu_lOGMiV:b-3p1dwh) ([raw file](https://github.com/robotlocomotion/drake/blob/0ee9497240193df2160dc86dd1f8fa4527a9aaf0/cmake/mex.cmake#L317)):*
> ```CMake
>       target_link_libraries(${target} liblast)
>     else()
>       set_target_properties(${target} PROPERTIES LINK_FLAGS "${MEXLIB_LDFLAGS}")
> ```

absent configuration-specific properties, does CMake set the debug level, etc flags itself?

---

*[drake/solvers/CMakeLists.txt, line 115 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3154#-KPtD13yx611FXAewJpN:-KPtD13yx611FXAewJpO:b-cyb36h) ([raw file](https://github.com/robotlocomotion/drake/blob/0ee9497240193df2160dc86dd1f8fa4527a9aaf0/drake/solvers/CMakeLists.txt#L115)):*
> ```CMake
> 
>   # These warnings are not worth fixing.
>   target_compile_options(NonlinearProgramSnoptmex PRIVATE
> ```

for my own education, why is this preferable?

---

*[drake/systems/plants/CMakeLists.txt, line 59 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3154#-KPridEnbisXCEMeMkbe:-KPtDAeXj92WBdcSqmiY:b-85i649) ([raw file](https://github.com/robotlocomotion/drake/blob/0ee9497240193df2160dc86dd1f8fa4527a9aaf0/drake/systems/plants/CMakeLists.txt#L59)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary>
> I see. I'd still prefer it in a separate commit (same PR is fine though!), or at least mention it in the commit message. (For that matter, I wouldn't mind a more detailed commit message in general...)
</details>
LGTM 

---

*[drake/systems/plants/constraint/CMakeLists.txt, line 20 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3154#-KPrimnMPO98UqLoEh-L:-KPtDDV201SMUhsuMyN9:bb6ywmx) ([raw file](https://github.com/robotlocomotion/drake/blob/0ee9497240193df2160dc86dd1f8fa4527a9aaf0/drake/systems/plants/constraint/CMakeLists.txt#L20)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary>
> Also, no.
</details>
LGTM

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3154#-:-KPt9irikns4tdZumRyI:brb94na)*

**Comment by jamiesnape** [comment]
Review status: 0 of 9 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[cmake/mex.cmake, line 317 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3154#-KPtCadmmq1pu_lOGMiU:-KPtF12nNxLLlIY5tWll:bylo3vv) ([raw file](https://github.com/robotlocomotion/drake/blob/0ee9497240193df2160dc86dd1f8fa4527a9aaf0/cmake/mex.cmake#L317)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> absent configuration-specific properties, does CMake set the debug level, etc flags itself?
</details>
Yes. That is main purpose of having CMAKE_BUILD_TYPE.

---

*[drake/solvers/CMakeLists.txt, line 115 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3154#-KPtD13yx611FXAewJpN:-KPtDXMZuCSBp4tqLApy:bm676ue) ([raw file](https://github.com/robotlocomotion/drake/blob/0ee9497240193df2160dc86dd1f8fa4527a9aaf0/drake/solvers/CMakeLists.txt#L115)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> for my own education, why is this preferable?
</details>
COMPILE_FLAGS is deprecated, but in addition this does not overwrite existing COMPILE_FLAGS.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3154)*

**Comment by jamiesnape** [comment]
Resolved merge conflicts.

**Comment by jamiesnape** [comment]
If anyone watching this thread wants to hit merge once the builds go green, feel free.

**Comment by david-german-tri** [comment]
:lgtm_strong: thanks!

---

Review status: 0 of 9 files reviewed at latest revision, 3 unresolved discussions.

---

*[cmake/mex.cmake, line 132 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3154#-KPtC1-BS5skBITs3gBV:-KPtV_xpOVx1YQa02_2o:b2c0zcb) ([raw file](https://github.com/robotlocomotion/drake/blob/0ee9497240193df2160dc86dd1f8fa4527a9aaf0/cmake/mex.cmake#L132)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> Is `get_mex_option` the "introspection" you've previously mentioned?  Could you add a comment to it explaining what it actually does?  (It looks like it assumes `mexv_output` is already in scope and parses it for these particular strings, or something like that?)
</details>
LGTM, thanks!

---

*[cmake/mex.cmake, line 231 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3154#-KPtCSr1oPtWeB4ZCINU:-KPtVop1fWusOdNCSAL_:b2c0zcb) ([raw file](https://github.com/robotlocomotion/drake/blob/0ee9497240193df2160dc86dd1f8fa4527a9aaf0/cmake/mex.cmake#L231)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> Meaning, the mex compiler suggested some flags, and we are editing them out so that CMake can set them instead?  Could you elaborate this comment a little?
</details>
LGTM, thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3154#-:-KPtW1HNZKyhQzo1KI2D:b-c5hpsr)*

**Comment by david-german-tri** [comment]
Reviewed 5 of 9 files at r1, 3 of 3 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3154)*

**Comment by david-german-tri** [comment]
@drake-jenkins-bot: retest this please

The linux_matlab failures look real, though.  :-/

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3154#-:-KPtbyUHA8ybxqn3fCET:b6wbq1e)*

**Comment by jamiesnape** [comment]
Apparently someone did a large amount of refactoring in the time it took to run the whole suite of MATLAB builds. Hopefully we are good now.

---

Review status: 8 of 9 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3154#-:-KPtxyo7YIVd1RurrWc6:b-raszwc)*
